pkg/hubble/peer/types/client.go

```diff
@@ -85,9 +77,7 @@ func (b RemoteClientBuilder) Client(target string) (Client, error) {
 		})
 		opts = append(opts, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
 	}
-	ctx, cancel := context.WithTimeout(context.Background(), b.DialTimeout)
```

Comment:
You can recreate this timeout by doing:
```
ctx, cancel := context.WithTimeout(...)
defer cancel()
conn.WaitForStateChange(ctx, connectivity.Ready)
```

and keep the dial timeout (as well as preventing spurious error messages). Seem reasonable?

Reply:
Sorry, this PR fell off my radar.

I agree the log change is a hack and probably not the correct approach, I just didn't want block this PR because of random warning logs.

But also IMO there isn't really a point in keeping this timeout, I doubt anyone ever configured this, as it just slightly changes the timing of reconnecting. And blocking dials are explicitly an anti-pattern (https://github.com/grpc/grpc-go/blob/master/Documentation/anti-patterns.md#especially-bad-using-deprecated-dialoptions), so don't really want to "re-implement" them.

Also `WaitForStateChange` is still marked as experimental, so I'm not sure it's significantly better to use that instead of the deprecated calls. And BTW it's not quite as simple because `WaitForStateChange` will only tell you when the state changed *from* the state you provide, no *to* the state..

IMO the PeerClient setup would need some refactoring to properly solve this where we:

1. Don't close the connection on any error. There is no point in doing that, gRPC will redial the connection if necessary. So only create the client once and reuse it
2. Do something smarter than just sleeping for 30 seconds on any failure and retrying. We should either use an exponential backoff, like we do for ClientConns, or see if we can introduce proper retries (https://github.com/grpc/grpc-go/blob/master/examples/features/retry/README.md)
3. Potentially implement some initial delay where we don't log errors. But this probably wouldn't be necessary when using proper gRPC retries.

But this is IMO out-of-scope for this PR, so I didn't want to propose it at first, but I might do it as a follow up myself.

Reply:
Hi,

For the third point, I have seen in the spire code that they check if the service they rely on is ready. Might be  better than a delay ? 

https://github.com/cilium/cilium/blob/fc86dfef89ae06478468fde7ca39786dd569b76d/operator/auth/spire/client.go#L317

I'll try this week to test it.

David

Reply:
Any progress on this?